Expose control of what part of text (start, middle or end) of line labels corresponds to the anchor position

nyalldawson opened the pull request:
Previously labels would always be centered over the anchor position of lines (unless it was a curved label, in which case it was the start of the lable!) This is limiting (and inconsistent) for some placements, where it is preferable to e.g. line up the end
of the label text with a specific point on the line.

There's also an option for "Follow placement", which means that the text anchor will depends on the placement along line setting:
- For labels anchored near the start of the line (0-25%), the anchor placement will be the **start** of the label text
- For labels anchored near the end of the line (75-100%), the anchor placement will be the **end** of the label text
- For anchor placements between 25 and 75%, the anchor placement will be the **center** of the label text

nyalldawson committed:
[feature][labeling] Expose control of what part of text (start,
middle or end) of line labels corresponds to the anchor position

Previously labels would always be centered over the anchor position
of lines (unless it was a curved label, in which case it was the
start of the lable!) This is limiting (and inconsistent) for some
placements, where it is preferable to e.g. line up the end
of the label text with a specific point on the line

elpaso commented on src/core/pal/feature.cpp:
```diff
@@ -1173,7 +1211,20 @@ std::size_t FeaturePart::createCandidatesAlongLineNearMidpoint( std::vector< std
+        textAnchorPoint = currentDistanceAlongLine;
+        break;
+      case QgsLabelLineSettings::AnchorTextPoint::CenterOfText:
+        textAnchorPoint = ( currentDistanceAlongLine + labelWidth / 2 );
```
```suggestion
        textAnchorPoint = currentDistanceAlongLine + labelWidth / 2;
```

elpaso reviewed (commented):
Nice work! 

Are you planning any tests?

DelazJ commented on src/core/labeling/qgslabelfeature.h:
```diff
@@ -471,6 +470,26 @@ class CORE_EXPORT QgsLabelFeature
     void setLineAnchorType( QgsLabelLineSettings::AnchorType type ) { mLineAnchorType = type; }
 
+    /**
+     * Returns the line anchor text point, which dictates which part of the label text
```
Does it need to be as this, mimicking the next function?
```suggestion
     * Returns the line anchor text \a point, which dictates which part of the label text
```

DelazJ commented on src/core/labeling/qgspallabeling.h:
```diff
@@ -336,6 +336,7 @@ class CORE_EXPORT QgsPalLayerSettings
       PolygonLabelOutside = 109, //!< Whether labels outside a polygon feature are permitted, or should be forced (since QGIS 3.14)
       LineAnchorPercent = 111, //!< Portion along line at which labels should be anchored (since QGIS 3.16)
       LineAnchorClipping = 112, //!< Clipping mode for line anchor calculation (since QGIS 3.20)
+      LineAnchorTextPoint = 115, //!< Line anchor text point (since QGIS 3.24)
```
```suggestion
      LineAnchorTextPoint = 115, //!< Line anchor text point (since QGIS 3.26)
```
?

nyalldawson committed:
Fix some labels disappear when set to strict label placement mode with curved placements

nirvn reviewed (approved):
Looks good.

nyalldawson committed:
Fix interaction between label overrun distances and strict placement modes

nyalldawson committed:
Use Q_ENUM for enums

nyalldawson committed:
Default to a "Follow placement" mode for the text anchor point, which
means when the line anchor is < 25% then it will be matched to the
start of the text, when the line anchor is > 75% it will be matched
to the end of the text, otherwise match to center of text

nyalldawson committed:
If in strict placement mode, we assume that labels can be overrun by whatever distance is required

nyalldawson committed:
Update some tests

nyalldawson committed:
Add tests

nyalldawson committed:
Fix test
